### PR TITLE
Adds pre-commit

### LINKED ISSUE #58: Add pre-commit with Ruff hook
### Feature description

Add [pre-commit](https://pre-commit.com/) to the repository to maintain code standards and formatting. This will remove the need for manual formatting and reduce code review times while improve code understandability. Current linting preference is [Ruff](https://docs.astral.sh/ruff/) due to performance and feature-set.

### Motivation

_No response_

### Possible implementation

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by BradyPlanden**
This PR adds the following (and closes #58): 

- Pre-commit to the repository
- Pre-commit CI job to the test_on_push workflow
- Updates CONTRIBUTING.MD for pre-commit workflow
- Runs the pre-commit hooks on the repo

### CODE REVIEW

File: .github/workflows/test_on_push.yaml [hunk 1]
```diff
@@ -14,6 +14,21 @@ concurrency:
...
+      - uses: actions/checkout@v4
+      - name: Setup Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.11
+
+      - name: Check formatting with pre-commit
+        run: |
+          python -m pip install pre-commit
+          pre-commit run ruff
```
Comment (martinjrobins) [inline .github/workflows/test_on_push.yaml:29]: nice, ruff is the shiny new toy :)

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        args: [--fix, --ignore=E741, --exclude=__init__.py]
+
+  - repo: https://github.com/nbQA-dev/nbQA
+    rev: 1.7.0
+    hooks:
+      - id: nbqa-ruff
+        additional_dependencies: [ruff==0.0.284]
+        args: ["--fix","--ignore=E501,E402"]
+
+  # - repo: https://github.com/adamchainz/blacken-docs
```
Comment (martinjrobins) [inline .pre-commit-config.yaml:19]: do we need this or can be deleted?

**Review by martinjrobins — APPROVED** [review]
looks great, thanks @BradyPlanden. Looking forward to using ruff (at least in the background)

File: .pre-commit-config.yaml (hunk 2 above)
Comment (BradyPlanden) [reply .pre-commit-config.yaml]: Agreed, I thought of leaving it for when we have the docs up and sorted, but we can just add it then.

**Pull request merged**